File: converters/TET-3.0-Linux/bind/python/fontfilter.py
#!/usr/bin/python
# Extract text from PDF and filter according to font name and size.
# This can be used to identify headings in the document and create a
# table of contents.
#
# $Id: fontfilter.py,v 1.3 2008/12/23 17:31:12 rjs Exp $
#/

import sys
from tetlib_py import *


# global option list */
globaloptlist = "searchpath={../data ../../../resource/cmap}"

# document-specific option list */
docoptlist = ""

# page-specific option list */
pageoptlist = "granularity=line"

# Search text with at least this size (use 0 to catch all sizes) */
fontsizetrigger = 10

# Catch text where the font name contains this string
# (use empty string to catch all font names)
#/
fontnametrigger = "Bold"

if len(sys.argv) != 2:
    raise Exception("usage: fontfilter <infilename>\n")

try:
    try:
	pageno = 0

	tet = TET_new()

	TET_set_option(tet, globaloptlist)

	doc = TET_open_document(tet, sys.argv[1], docoptlist)

	if (doc == -1):
	    raise Exception("Error %d in %s(): %s\n" % \
		(TET_get_errnum(tet), TET_get_apiname(tet), \
		TET_get_errmsg(tet)))

	# get number of pages in the document */
	n_pages = int(TET_pcos_get_number(tet, doc, "length:pages"))

	# loop over pages in the document */
	for pageno in range(1, n_pages+1):
	    page = TET_open_page(tet, doc, pageno, pageoptlist)

	    if (page == -1):
		print ("Error %d in %s() on page %d: %s\n" % \
		    (TET_get_errnum(tet), TET_get_apiname(tet), pageno, \
		     TET_get_errmsg(tet)))
		continue                        # try next page */

	    # Retrieve all text fragments for the page */
	    text = TET_get_text(tet, page)
	    while (text):
		# Loop over all characters */
		ci = TET_get_char_info(tet, page)
		while (ci):
		    # We need only the font name and size the text 
		    # position could be fetched from ci->x and ci->y.
		    #/
		    fontname = TET_pcos_get_string(tet, doc, \
				"fonts[%d]/name" % ci["fontid"])

		    # Check whether we found a match */
		    # C only: some versions of strstr don't allow empty
		    # strings, so we better check */
		    if (ci["fontsize"] >= fontsizetrigger and \
				fontname.find(fontnametrigger) != -1):
			# print the retrieved font name, size, and text */
			print ("[%s %.2f] %s" % \
			    (fontname, ci["fontsize"], text))
		    ci = TET_get_char_info(tet, page)
		    # In this sample we check only the first character of
		    # each fragment.
		    #/
		    break
		text = TET_get_text(tet, page)
	    if (TET_get_errnum(tet) != 0):
		raise Exception("Error %d in %s() on page %d: %s\n" % \
		    (TET_get_errnum(tet), TET_get_apiname(tet), pageno, \
		     TET_get_errmsg(tet)))

	    TET_close_page(tet, page)

	TET_close_document(tet, doc)


    except TETlibException:
	if (pageno == 0):
	    print("\nError " + repr(TET_get_errnum(tet)) + "in " + \
		TET_get_apiname(tet) + ": " + TET_get_errmsg(tet) + "\n")
	else:
	    print("\nError " + repr(TET_get_errnum(tet)) + "in " + \
		TET_get_apiname(tet) + "() on page " + repr(pageno) \
		+ ": " + TET_get_errmsg(tet) + "\n")

    except Exception, e:
	print "Exception occurred:"
	print e

finally:
    TET_delete(tet)

File: converters/TET-3.0-Linux/bind/perl/fontfilter.pl
#!/usr/bin/perl
# Extract text from PDF and filter according to font name and size.
# This can be used to identify headings in the document and create a
# table of contents.
#
# $Id: fontfilter.pl,v 1.3 2008/12/23 17:31:11 rjs Exp $
#/

use tetlib_pl 3.0;
use strict;

# global option list */
my $globaloptlist = "searchpath={../data ../../../resource/cmap}";

# document-specific option list */
my $docoptlist = "";

# page-specific option list */
my $pageoptlist = "granularity=line";

# Search text with at least this size (use 0 to catch all sizes) */
my $fontsizetrigger = 10;

# Catch text where the font name contains this string
# (use empty string to catch all font names)
#/
my $fontnametrigger = "Bold";
my $pageno = 0;

eval  {
    my $tet;
    my $n_pages;
    my $doc;

    if ($#ARGV != 0) {
	die("usage: fontfilter <infilename>\n");
    }

    $tet = TET_new();

    TET_set_option($tet, $globaloptlist);

    $doc = TET_open_document($tet, $ARGV[0], $docoptlist);

    if ($doc == -1) {
	die(sprintf("Error %d in %s(): %s\n",
	    TET_get_errnum($tet), TET_get_apiname($tet), TET_get_errmsg($tet)));
    }

    # get number of pages in the document */
    $n_pages = TET_pcos_get_number($tet, $doc, "length:pages");

    # loop over pages in the document */
    for ($pageno = 1; $pageno <= $n_pages; ++$pageno) {
	my $text;
	my $page;

	$page = TET_open_page($tet, $doc, $pageno, $pageoptlist);

	if ($page == -1) {
	    printf("Error %d in %s() on page %d: %s\n",
		TET_get_errnum($tet), TET_get_apiname($tet), $pageno,
		TET_get_errmsg($tet));
	    next;                        # try next page */
	}

	# Retrieve all text fragments for the page */
	while ($text = TET_get_text($tet, $page)) {
	    my $ci;
	    my $fontname;

	    # Loop over all characters */
	    while ($ci = TET_get_char_info($tet, $page)) {
		# We need only the font name and size; the text 
		# position could be fetched from ci->x and ci->y.
		#/
		$fontname = TET_pcos_get_string($tet, $doc,
			    "fonts[" . $ci->{"fontid"} . "]/name");

		# Check whether we found a match */
		if ($ci->{"fontsize"} >= $fontsizetrigger &&
			($fontname =~ m/$fontnametrigger/)) {
		    # print the retrieved font name, size, and text */
		    printf("[%s %.2f] %s\n",
			$fontname, $ci->{"fontsize"}, $text);
		}

		# In this sample we check only the first character of
		# each fragment.
		#/
		last;
	    }
	}

	if (TET_get_errnum($tet) != 0) {
	    printf("Error %d in %s() on page %d: %s\n",
		TET_get_errnum($tet), TET_get_apiname($tet), $pageno,
		TET_get_errmsg($tet));
	}

	TET_close_page($tet, $page);
    }

    TET_close_document($tet, $doc);
};

if ($@) {
    printf("PDFlib Exception occurred:\n");
    if ($pageno == 0) {
	printf("Error $@\n");
    } else {
	printf("Error $@ on page $pageno\n");
    }
    exit;
}
